File: apps/frontend/src/types/ticket.ts
export interface AttachmentFile {
  id: string
  s3_key: string
  filename: string
  size_bytes: number
  mime_type: string
  url?: string
  uploaded_at: string
}

// ── Legacy MaintenanceTicket (meter discrepancy from inspection flow) ──────────

export type LegacyTicketStatus = 'open' | 'resolved'

export interface MaintenanceTicket {
  id: string
  org_id: string
  property_id: string
  unit_id: string
  lease_id: string
  inspection_report_id: string
  ticket_type: string
  title: string
  description: string
  utility_key: string
  utility_label: string
  system_reading: number | null
  reported_reading: number
  status: LegacyTicketStatus
  assigned_to: string | null
  resolution_reading: number | null
  resolution_notes: string | null
  evidence_urls: string[]
  resolved_by: string | null
  resolved_at: string | null
  created_at: string
  updated_at: string
}

// ── New comprehensive Ticket ──────────────────────────────────────────────────

export type TicketStatus =
  | 'open'
  | 'assigned'
  | 'in_progress'
  | 'pending_review'
  | 'resolved'
  | 'closed'
  | 'cancelled'

export type TicketPriority = 'low' | 'normal' | 'high' | 'urgent'

export type TicketCategory =
  | 'maintenance'
  | 'utility_reading'
  | 'move_in_inspection'
  | 'move_out_request'
  | 'move_out_inspection'
  | 'move_out_refund'
  | 'request'
  | 'complaint'
  | 'other'

export type TicketTaskType = 'meter_reading' | 'inspection_item' | 'checklist_item' | 'custom'
export type TicketTaskStatus = 'pending' | 'in_progress' | 'completed' | 'skipped'
export type TicketTaskCondition = 'good' | 'fair' | 'poor' | 'damaged'

export interface TicketTask {
  id: string
  title: string
  task_type: TicketTaskType
  status: TicketTaskStatus
  meter_number?: string
  previous_reading?: number
  current_reading?: number
  unit_of_measure?: string
  room?: string
  condition?: TicketTaskCondition
  notes?: string
  attachment_urls: string[]
  assigned_to?: string
  // billing linkage (meter_reading tasks)
  unit_id?: string
  unit_code?: string
  tenant_name?: string
  invoice_id?: string
  line_item_id?: string
  utility_key?: string
  body_html?: string
  attachments?: AttachmentFile[]
  completed_at?: string
  created_at: string
  updated_at: string
}

export interface OrgMember {
  id: string
  first_name: string
  last_name: string
  email: string
  role: string
}

export interface TicketComment {
  id: string
  author_id: string
  author_role: string
  author_name?: string
  body: string
  body_html?: string
  attachments?: AttachmentFile[]
  attachment_urls: string[]
  created_at: string
}

export interface TicketActivity {
  id: string
  type: string
  actor_id?: string
  actor_role?: string
  actor_name?: string
  description: string
  created_at: string
}

export interface Ticket {
  id: string
  reference_number?: string
  org_id: string
  property_id: string
  property_name?: string
  unit_id?: string
  unit_label?: string
  tenant_id?: string
  tenant_name?: string
  assigned_to?: string
  assigned_to_name?: string
  creator_id: string
  creator_name?: string
  creator_role?: string
  category: TicketCategory | string
  priority: TicketPriority
  status: TicketStatus
  title: string
  description?: string
  body_html?: string
  attachments?: AttachmentFile[]
  attachment_urls: string[]
  comments: TicketComment[]
  activity: TicketActivity[]
  tasks: TicketTask[]
  submission_token?: string
  submission_data?: Record<string, unknown>
  submitted_at?: string
  resolution_notes?: string
  resolved_at?: string
  closed_at?: string
  capture_started_at?: string
  capture_completed_at?: string
  created_by: string
  created_at: string
  updated_at: string
}

export interface TicketListResponse {
  items: Ticket[]
  total: number
  page: number
  page_size: number
}

export interface TicketCounts {
  open: number
  assigned: number
  in_progress: number
  pending_review: number
  resolved: number
  closed: number
  cancelled: number
  total: number
}

export interface TicketCreatePayload {
  property_id: string
  unit_id?: string
  category: string
  priority?: string
  title: string
  description?: string
  body_html?: string
  tenant_id?: string
}

export interface TicketUpdatePayload {
  status?: string
  assigned_to?: string
  priority?: string
  title?: string
  description?: string
  resolution_notes?: string
}

export interface BulkUtilityTicketPayload {
  property_id: string
  unit_ids: string[]
  title: string
  description?: string
}

export interface TicketTaskCreatePayload {
  title: string
  task_type?: string
  meter_number?: string
  previous_reading?: number
  unit_of_measure?: string
  room?: string
  notes?: string
  assigned_to?: string
}

export interface TicketTaskUpdatePayload {
  status?: string
  title?: string
  meter_number?: string
  previous_reading?: number
  current_reading?: number
  unit_of_measure?: string
  room?: string
  condition?: string
  notes?: string
  assigned_to?: string
}

File: apps/backend/app/schemas/ticket.py
"""Ticket schemas — covers both MaintenanceTicket (legacy) and new comprehensive Ticket."""
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ── Legacy MaintenanceTicket schemas (meter discrepancy) ──────────────────────

class TicketResolveRequest(BaseModel):
    resolution_reading: float = Field(gt=0, description="The official move-in meter reading adopted")
    resolution_notes: Optional[str] = None


class TicketResponse(BaseModel):
    id: str
    org_id: str
    property_id: str
    unit_id: str
    lease_id: str
    inspection_report_id: str
    ticket_type: str
    title: str
    description: str
    utility_key: str
    utility_label: str
    system_reading: Optional[float]
    reported_reading: float
    status: str
    assigned_to: Optional[str]
    resolution_reading: Optional[float]
    resolution_notes: Optional[str]
    evidence_urls: List[str] = []
    resolved_by: Optional[str]
    resolved_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime


# ── New comprehensive Ticket schemas ──────────────────────────────────────────

class GeneralTicketCreateRequest(BaseModel):
    property_id: str
    unit_id: Optional[str] = None
    category: str
    priority: str = "normal"
    title: str
    description: Optional[str] = None
    tenant_id: Optional[str] = None
    org_id: Optional[str] = None


class GeneralTicketUpdateRequest(BaseModel):
    status: Optional[str] = None
    assigned_to: Optional[str] = None
    priority: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    resolution_notes: Optional[str] = None


class TicketCommentRequest(BaseModel):
    body: str


class BulkUtilityTicketRequest(BaseModel):
    property_id: str
    unit_ids: List[str]
    title: str
    description: Optional[str] = None


class SubmissionDataRequest(BaseModel):
    data: Dict[str, Any]


# ── Task schemas ──────────────────────────────────────────────────────────────

class TicketTaskCreateRequest(BaseModel):
    title: str
    task_type: str = "custom"   # meter_reading | inspection_item | checklist_item | custom
    meter_number: Optional[str] = None
    previous_reading: Optional[float] = None
    unit_of_measure: Optional[str] = "units"
    room: Optional[str] = None
    notes: Optional[str] = None
    assigned_to: Optional[str] = None


class TicketTaskUpdateRequest(BaseModel):
    status: Optional[str] = None          # pending|in_progress|completed|skipped
    title: Optional[str] = None
    meter_number: Optional[str] = None
    previous_reading: Optional[float] = None
    current_reading: Optional[float] = None
    unit_of_measure: Optional[str] = None
    room: Optional[str] = None
    condition: Optional[str] = None       # good|fair|poor|damaged
    notes: Optional[str] = None
    assigned_to: Optional[str] = None


class TicketTaskResponse(BaseModel):
    id: str
    title: str
    task_type: str
    status: str
    meter_number: Optional[str] = None
    previous_reading: Optional[float] = None
    current_reading: Optional[float] = None
    unit_of_measure: Optional[str] = None
    room: Optional[str] = None
    condition: Optional[str] = None
    notes: Optional[str] = None
    attachment_urls: List[str] = []
    assigned_to: Optional[str] = None
    # billing linkage fields (present on meter_reading tasks)
    unit_id: Optional[str] = None
    unit_code: Optional[str] = None
    tenant_name: Optional[str] = None
    invoice_id: Optional[str] = None
    line_item_id: Optional[str] = None
    utility_key: Optional[str] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class TaskPublicUpdateRequest(BaseModel):
    current_reading: float
    notes: Optional[str] = None
    captured_by: Optional[str] = None  # user_id passed via query param
    photo_key: Optional[str] = None    # S3 key from prior photo upload
    meter_number: Optional[str] = None # meter identifier (sets on unit if not already set)


class MeterReadResponse(BaseModel):
    reading: Optional[float] = None
    confidence: float = 0.0
    raw_text: Optional[str] = None
    error: Optional[str] = None


# ── Enriched response schemas ─────────────────────────────────────────────────

class TicketCommentResponse(BaseModel):
    id: str
    author_id: str
    author_role: str
    author_name: Optional[str] = None
    body: str
    attachment_urls: List[str] = []
    created_at: datetime


class TicketActivityResponse(BaseModel):
    id: str
    type: str
    actor_id: Optional[str] = None
    actor_role: Optional[str] = None
    actor_name: Optional[str] = None
    description: str
    created_at: datetime


class OrgMemberResponse(BaseModel):
    """Lightweight user record for assignment dropdowns."""
    id: str
    first_name: str
    last_name: str
    email: str
    role: str

    @property
    def full_name(self) -> str:
        return f"{self.first_name} {self.last_name}".strip()


class GeneralTicketResponse(BaseModel):
    id: str
    reference_number: Optional[str] = None
    org_id: str
    property_id: str
    property_name: Optional[str] = None   # resolved
    unit_id: Optional[str] = None
    unit_label: Optional[str] = None      # resolved (unit_code / unit_number)
    tenant_id: Optional[str] = None
    tenant_name: Optional[str] = None     # resolved
    assigned_to: Optional[str] = None
    assigned_to_name: Optional[str] = None  # resolved
    creator_id: Optional[str] = None  # resolved
    creator_name: Optional[str] = None    # resolved
    creator_role: Optional[str] = None
    category: str
    priority: str
    status: str
    title: str
    description: Optional[str] = None
    attachment_urls: List[str] = []
    comments: List[TicketCommentResponse] = []
    activity: List[TicketActivityResponse] = []
    tasks: List[TicketTaskResponse] = []
    submission_token: Optional[str] = None
    submission_data: Optional[Dict[str, Any]] = None
    submitted_at: Optional[datetime] = None
    resolution_notes: Optional[str] = None
    resolved_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    capture_started_at: Optional[datetime] = None
    capture_completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class GeneralTicketListResponse(BaseModel):
    items: List[GeneralTicketResponse]
    total: int
    page: int
    page_size: int


class TicketCountsResponse(BaseModel):
    open: int = 0
    assigned: int = 0
    in_progress: int = 0
    pending_review: int = 0
    resolved: int = 0
    closed: int = 0
    cancelled: int = 0
    total: int = 0
